websocket/connection.go

```diff
@@ -78,38 +76,44 @@ type ManagedConnection struct {
 // that can only send messages to the endpoint it connects to.
 // The connection will continuously be kept alive and reconnected
 // in case of a loss of connectivity.
-func NewDurableSendingConnection(target string) *ManagedConnection {
-	return NewDurableConnection(target, nil)
+func NewDurableSendingConnection(target string, logger *zap.SugaredLogger) *ManagedConnection {
+	return NewDurableConnection(target, nil, logger)
 }
 
 // NewDurableConnection creates a new websocket connection, that
 // passes incoming messages to the given message channel. It can also
 // send messages to the endpoint it connects to.
 // The connection will continuously be kept alive and reconnected
 // in case of a loss of connectivity.
-func NewDurableConnection(target string, messageChan chan []byte) *ManagedConnection {
-	c := newConnection(target, messageChan)
+func NewDurableConnection(target string, messageChan chan []byte, logger *zap.SugaredLogger) *ManagedConnection {
+	websocketConnectionFactory := func(target string) (rawConnection, error) {
+		dialer := &websocket.Dialer{
+			HandshakeTimeout: 3 * time.Second,
+		}
+		conn, _, err := dialer.Dial(target, nil)
+		return conn, err
+	}
+
+	c := newConnection(target, messageChan, websocketConnectionFactory)
 
 	// Keep the connection alive asynchronously and reconnect on
 	// connection failure.
 	go func() {
 		// If the close signal races the connection attempt, make
 		// sure the connection actually closes.
-		defer func() {
-			c.connectionLock.RLock()
-			defer c.connectionLock.RUnlock()
-
-			if conn := c.connection; conn != nil {
-				conn.Close()
-			}
-		}()
+		defer c.Shutdown()
 		for {
 			select {
 			default:
+				logger.Infof("Connecting to %q", target)
 				if err := c.connect(); err != nil {
+					logger.Errorw(fmt.Sprintf("Connecting to %q failed", target), zap.Error(err))
 					continue
 				}
-				c.keepalive()
+				logger.Infof("Connected to %q", target)
+				if err := c.keepalive(); err != nil {
```

Comment:
This in theory never ends, thus `c.closeChan` never gets the chance to be read below.
Is this what we intend?

Reply:
It isn't, added a "close" breaker here as well. The thinking was though, that a call to `Shutdown` closes the connection which would then cause keepalive to error out.